File: campaign-os/_lib/visibility_guard.py
# Step 97/98 Visibility Guard — Python language adapter.
# Mirrors scripts/_lib/visibility-guard.js. JS is canonical. Same inputs,
# same outputs, same locked behaviour. No I/O, no persistence, no subprocess.

VALID_OPERATOR_STATES = ('visible', 'not-visible', 'disputed', 'unknown')
BLOCKING_STATES = {'VISIBILITY_DISPUTED', 'EXTERNAL_STATE_DISPUTED'}
STATE_CHANGING_ACTIONS = frozenset({
    'reconcile', 'delete', 'duplicate-skip',
    'mark-live', 'destructive-state-reconciliation',
})


def assert_no_visibility_dispute(apiState, canonicalState, operatorVisibilityState):
    # Treat None / '' as "no value supplied" -> default to 'unknown'.
    # Treat any other non-valid state as invalid input -> fail closed.
    op_provided = operatorVisibilityState is not None and operatorVisibilityState != ''
    if op_provided and operatorVisibilityState not in VALID_OPERATOR_STATES:
        return {'state': 'VISIBILITY_DISPUTED', 'allowStateChange': False,
                'reason': f'invalid operator state: {operatorVisibilityState}'}
    op_state = operatorVisibilityState if op_provided and operatorVisibilityState in VALID_OPERATOR_STATES else 'unknown'
    if op_state in ('not-visible', 'disputed'):
        return {'state': 'VISIBILITY_DISPUTED', 'allowStateChange': False,
                'reason': f'operator reports {operatorVisibilityState}'}
    if not canonicalState or canonicalState == 'missing':
        return {'state': 'NO_OBJECT', 'allowStateChange': False,
                'reason': 'no canonical object'}
    if not apiState or apiState == 'missing':
        return {'state': 'EXTERNAL_STATE_DISPUTED', 'allowStateChange': False,
                'reason': 'canonical exists but API missing'}
    return {'state': 'OK', 'allowStateChange': True,
            'reason': 'canonical and API agree'}


def blocks_action(guard, action):
    return guard['state'] in BLOCKING_STATES and action in STATE_CHANGING_ACTIONS

File: scripts/_lib/visibility-guard.js
// Step 97/98 Visibility Guard — canonical JavaScript implementation.
// Pure function. No I/O, no logging, no persistence. Caller passes
// operatorVisibilityState (defaults to 'unknown'). Disputes are sticky at
// the caller; this module never records state. Mirrored in Python at
// campaign-os/_lib/visibility_guard.py; JS is canonical.

'use strict';

const VALID_OPERATOR_STATES = ['visible', 'not-visible', 'disputed', 'unknown'];
const BLOCKING_STATES = new Set(['VISIBILITY_DISPUTED', 'EXTERNAL_STATE_DISPUTED']);
// "Missing-like" values for apiState / canonicalState: empty/undefined/null + 'missing' + 'absent'.
// All are treated as "not present". Only truthy non-missing values (e.g. 'exists') count as present.
const ABSENT_LIKE = v => v === undefined || v === null || v === '' || v === 'missing' || v === 'absent';

const STATE_CHANGING_ACTIONS = [
  'reconcile',
  'delete',
  'duplicate-skip',
  'mark-live',
  'destructive-state-reconciliation',
];

function assertNoVisibilityDispute(context) {
  if (!context || typeof context !== 'object') {
    return { state: 'NO_OBJECT', allowStateChange: false, reason: 'invalid context' };
  }
  const { apiState, canonicalState, operatorVisibilityState } = context;
  // Operator state: missing-like -> 'unknown' (no opinion). Any other invalid value fails closed.
  if (!ABSENT_LIKE(operatorVisibilityState) && !VALID_OPERATOR_STATES.includes(operatorVisibilityState)) {
    return { state: 'VISIBILITY_DISPUTED', allowStateChange: false, reason: `invalid operator state: ${operatorVisibilityState}` };
  }
  const opState = ABSENT_LIKE(operatorVisibilityState) ? 'unknown' : operatorVisibilityState;
  if (opState === 'not-visible' || opState === 'disputed') {
    return { state: 'VISIBILITY_DISPUTED', allowStateChange: false, reason: `operator reports ${operatorVisibilityState}` };
  }
  // Canonical is the authority. Canonical missing + API missing -> NO_OBJECT.
  const canonicalMissing = ABSENT_LIKE(canonicalState);
  const apiMissing = ABSENT_LIKE(apiState);
  if (canonicalMissing && apiMissing) {
    return { state: 'NO_OBJECT', allowStateChange: false, reason: 'no canonical object' };
  }
  // Canonical exists but API does not -> EXTERNAL_STATE_DISPUTED.
  if (!canonicalMissing && apiMissing) {
    return { state: 'EXTERNAL_STATE_DISPUTED', allowStateChange: false, reason: 'canonical exists but API missing' };
  }
  // Canonical missing but API exists -> API_ONLY (canonical is authority: nothing to change here,
  // but flag the discrepancy). No operator dispute -> do NOT block.
  if (canonicalMissing && !apiMissing) {
    return { state: 'API_ONLY', allowStateChange: true, reason: 'api confirms existence; canonical has no record' };
  }
  return { state: 'OK', allowStateChange: true, reason: 'canonical and API agree' };
}

function blocksAction(guard, action) {
  if (!BLOCKING_STATES.has(guard.state)) return false;
  return STATE_CHANGING_ACTIONS.includes(action);
}

module.exports = {
  assertNoVisibilityDispute,
  blocksAction,
  VALID_OPERATOR_STATES,
  BLOCKING_STATES,
  STATE_CHANGING_ACTIONS,
};